File: c/problem_15.c
#include "problem_15.h"
#include <stdio.h>
#include <assert.h>

/* reservoirSample: O(n) time, O(1) space. Uses LCG for reproducibility. */
int reservoirSample(const int *stream, int n, unsigned int seed) {
    unsigned int s = seed;
    int result = stream[0];
    for (int i = 1; i < n; i++) {
        s = s * 1664525u + 1013904223u;
        unsigned int r = (s >> 1) % (unsigned int)(i + 1);
        if (r == 0) result = stream[i];
    }
    return result;
}

void problem15(void) {
    printf("Problem 15:\n");
    int stream[10] = {1,2,3,4,5,6,7,8,9,10};
    int counts[11] = {0};
    int trials = 100000;
    for (int seed = 0; seed < trials; seed++)
        counts[reservoirSample(stream, 10, (unsigned int)seed)]++;
    for (int v = 1; v <= 10; v++) {
        double freq = counts[v] / (double)trials;
        assert(freq >= 0.07 && freq <= 0.13);
    }
    printf("All Tests Passed\n");
}

File: rust/problem_15.rs
// reservoir_sample picks one element uniformly at random from a stream.
// Uses a simple LCG for reproducibility. Runs in O(n) time and O(1) space.
pub fn reservoir_sample(stream: &[i64], seed: u64) -> i64 {
    let mut s = seed;
    let mut lcg_next = |n: u64| -> u64 {
        s = s.wrapping_mul(6364136223846793005).wrapping_add(1442695040888963407);
        (s >> 33) % n
    };
    let mut result = stream[0];
    for i in 1..stream.len() {
        if lcg_next(i as u64 + 1) == 0 { // probability 1/(i+1)
            result = stream[i];
        }
    }
    result
}

pub fn problem15() {
    println!("Problem 15:");
    let stream: Vec<i64> = (1..=10).collect();
    let mut counts = std::collections::HashMap::<i64, u64>::new();
    let trials: u64 = 100_000;
    for seed in 0..trials {
        let v = reservoir_sample(&stream, seed);
        *counts.entry(v).or_insert(0) += 1;
    }
    for &v in &stream {
        let freq = *counts.get(&v).unwrap_or(&0) as f64 / trials as f64;
        assert!(freq >= 0.08 && freq <= 0.12,
            "assertion failed: element {} frequency {}", v, freq);
    }
    println!("All Tests Passed");
}
